3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,253 @@
+            .receiveEventMessage(event);
+        });
+        this.flushQueues_();
+      });
+    this.sendReadyMessageToCreative_();
+  }
+
+  /**
+   * Gets a handle to the window (or testing object that mocks window)
+   * @returns {!Window}
+   */
+  getWindow() {
+    return this.win_;
+  }
+
+  /**
+   * Sends a message from the third-party vendor's metrics-collection page back
+   * to the creative.
+   * @private
+   */
+  sendReadyMessageToCreative_() {
+    this.iframeMessagingClient_.sendMessage(AMP_ANALYTICS_3P_MESSAGE_TYPE.READY,
+      {});
+  }
+
+  /**
+   * Sends a message back to the host window
+   * @param {string} type The type of message to send.
+   * @param {(ampAnalytics3pReadyMessage|ampAnalytics3pResponse)} opt_payload
+   * The payload of message to send.
+   */
+  sendMessage(type, opt_payload) {
+    this.iframeMessagingClient_.sendMessage(type, opt_payload);
+  }
+
+  /**
+   * Instructs each AmpAnalytics3pCreativeMessageRouter instances to send its
+   * queued messages to its listener (if if has a listener yet).
+   * @private
+   */
+  flushQueues_() {
+    Object.entries(this.creativeMessageRouters_).forEach(
+      entry => {
+        try {
+          const creativeMessageRouter = entry[1];
+          creativeMessageRouter.sendQueuedMessagesToListener();
+        } catch (e) {
+          dev().error(TAG_, 'Caught exception executing listener for ' +
+            entry[0] + ': ' + e.message);
+        }
+      });
+  }
+}
+
+new AmpAnalytics3pMessageRouter(window);
+
+/**
+ * Receives messages bound for this cross-domain iframe, from a particular
+ * creative
+ */
+class AmpAnalytics3pCreativeMessageRouter {
+  /**
+   * @param {!AmpAnalytics3pMessageRouter} parent
+   * @param {!string} creativeId
+   */
+  constructor(parent, creativeId) {
+    /** @private {!AmpAnalytics3pMessageRouter} */
+    this.parent_ = parent;
+
+    /**
+     * @private {string}
+     */
+    this.creativeId_ = creativeId;
+
+    /**
+     * @private {string}
+     */
+    this.extraData_ = null;
+
+    /**
+     * private {!Array<ampAnalytics3pEvent>}
+     */
+    this.receivedMessagesQueue_ = [];
+
+    /**
+     * private {function(Array<ampAnalytics3pEvent>)=}
+     */
+    this.eventListener_ = null;
+
+    if (this.parent_.getWindow().onNewAmpAnalyticsInstance) {
+      this.parent_.getWindow().onNewAmpAnalyticsInstance(this);
+    } else {
+      dev().error(TAG_, 'Vendor page must implement onNewAmpAnalyticsInstance' +
+        ' prior to loading library script.');
+    }
+  }
+
+  /**
+   * Registers a callback function to be called when AMP Analytics events occur.
+   * There may only be one listener. If another function has previously been
+   * registered as a listener, it will no longer receive events.
+   * @param {!function(!Array<ampAnalytics3pEvent>)} listener A function
+   * that takes an array of event strings, and does something with them.
+   */
+  registerAmpAnalytics3pEventsListener(listener) {
```

Comment:
Let's make it clear if its called twice:
if (this.eventListener_) {
   user().warn('Replacing existing eventListener for....');
}